### ISSUE TITLE
Reports (day, week) differ from view in App

### ISSUE THREAD

**Issue by froggi123** [mathisdt/trackworktime#100]
To me it looks like there are some wrong calculation when generating Reports. By manually calculating from event export I get the same as the app display on screen.
The daily report differs in 3 Minutes in my example on Monday. And each day has a little difference. Finally in the week report there are 9 Minutes missing.

For reference I attach an LibreOffice document, with screenshot and copied Report export.

[TWT_Report_wrong_export.zip](https://github.com/mathisdt/trackworktime/files/4743069/TWT_Report_wrong_export.zip)

**Comment by froggi123** [comment]
Forgot to mention, I'm using Version 1.0.6

**Comment by mathisdt** [comment]
@froggi123 Thanks for reporting, the report calculation indeed was not exact. But much has happened since, the app is at v1.1.8 currently and has a reworked calculation engine. Could you please check if the error still exists? I'm optimistic that it's fixed now.

**Comment by Bawwr** [comment]
Hey , im on 1.1.9 and the issue still exists.
The app shows a work time for the whole week of 43:31
when i export to sums per day current week i get a worktime of 43:28

**Comment by mathisdt** [comment]
@froggi123 / @Bawwr I still can't reproduce it, could you please verify this issue still exists with the current release of TWT (1.1.12)?

If it does, please provide the following data so I can debug this:

1. screenshot of main table
2. screenshot(s) of event list (shown if you tap on the right part of the main table - if it doesn't fit on one screen, please provide multiple images)
3. the export(s) which show a differing time amount

When I have the data, I can look into the code to identify the point where the app does the wrong thing, but at the moment I'm stuck because in my tests all seems fine.

If you don't want to post your data here, you can also [email it to me](https://zephyrsoft.org/contact-about-me).

**Comment by tuxayo** [comment]
Reproduce:
1. Track time for 15 to 30 sec, pause and repeat until a few minutes in total
2. reports: current, week, events times per task weekly
3. Preview
4. Preview shows 0 min

With real life data, I have examples of past weeks where the differences are
16:26 in app vs 15:24 in reports
5:24 in app vs 5:07 in reports

**Comment by tuxayo** [comment]
Great news, thanks a lot :D
I'm soon starting to invoice work based on time so the app loosing time would have been bad ^^"

> When testing in the past, I created events from the dialog and not by clocking in and out using the buttons. But only "real" events (which were not entered via the dialog) create the problem here as they had also the seconds saved with them

That's what I meant in #209 by «events put via the start and stop button, not manually. Because they need to not be round to the minute. I suspect it's a truncation issue.»

> I'm somewhat hesitant to push this to all users too fast because the time calculation is kind of the "heart" of the app.

I can help testing, here is what I indent to do:

Track time for 15 to 30 sec, pause (sometimes shortly sometime for more than 1min) and repeat until a few minutes in total. Then again with a different task. Do it in parallel with a stopwatch. And compare that the stopwatch matches the in app total and the reports.

**Comment by mathisdt** [comment]
> Track time for 15 to 30 sec, pause (sometimes shortly sometime for more than 1min) and repeat until a few minutes in total. Then again with a different task. Do it in parallel with a stopwatch. And compare that the stopwatch matches the in app total and the reports.

I think comparing the results of TWT with a stopwatch would fail. TWT is only displaying times with hour and minute, and my fix was to cut off the seconds when clocking in or out via the button, so the simpler sum calculation from the reports now should match the results shown on the main screen. My base assumption was that everything less than a minute doesn't have to count - or, to be more specific, everything not having at least a change of minute between start and end.

**Comment by tuxayo** [comment]
> my fix was to cut off the seconds when clocking in or out via the button, so the simpler sum calculation from the reports now should match the results shown on the main screen

That moves the problem. TWT still loses time. I tend to loose focus easily so I have a lot of periods between 1 min (event less) and 10 min. Is seconds are always truncated, that's 30 sec of time lost on average per period. So it adds up to something significant. Isn't there a way to not loss the seconds?

**Comment by mathisdt** [comment]
Theoretically the app could also display all times with hours, minutes and seconds, but firstly this would be a lot of work (plus the editing possibilities would have to be extended) and secondly I think this would really clutter the UI which is already not the cleanest.

If the app doesn't display also the seconds, it shouldn't take them into account for calculation because the user can't see neither edit them.

@tuxayo If you know a way to solve this, please share...

**Comment by tuxayo** [comment]
@mathisdt 
> Theoretically the app could also display all times with hours, minutes and seconds, but firstly this would be a lot of work (plus the editing possibilities would have to be extended) and secondly I think this would really clutter the UI which is already not the cleanest.

Indeed, displaying seconds is no needed.

> If the app doesn't display also the seconds, it shouldn't take them into account for calculation because the user can't see neither edit them.

That shouldn't be linked. We don't want the seconds displayed but we should want them accounted to not loose non negligible amounts of time. Having them accounted by the start and stop button and not lost by the in app total and the reports is enough to fix the issue. No need to have them accounted in the UI to edit events.

Having TWT to have a negligible difference vs a stopwatch is important to be able to rely on it.

**Comment by mathisdt** [comment]
I suspect that most users will interpret the loss of less than one minute as negligible. Your use case seems special to me, the "normal users" will clock their 6 to 10 hours a day with the app and probably won't care about 30 seconds. Plus, the automatic clock-in / clock-out by location and by Wi-Fi is also checking only once per minute if an event should be generated. It's all planned to work "minute-wise".

Should it be changed, then it has to be at least behind a switch in the settings (which is disabled by default) so the users who don't care about less than a minute won't be puzzled.

**Comment by tuxayo** [comment]
I too interpret the loss of less than one minute as negligible whether it's for in app total or per task total in reports. In the end it's about 1 hours lost over 16 hours over 120 periods [1]. That's 8 minutes on average due to juggling over tasks, loosing focus, being interrupted, water break, pee break, snack breaks.

Even for other users, it's not less than one minute in total, it's 30 sec per period.

[1] This matches perfectly with an average of 30 sec lost when truncating second on a random time. 120*30 sec = on hour.

**Comment by tuxayo** [comment]
On way to fix this would be to not have the last change that adds imprecision to the start and stop buttons and to have the reports not truncate the seconds. It might require to change int to floats for the minute vars in the report calculations.

Another way is to not have the last change that adds imprecision to the start and stop buttons and to have the reports round to the minutes instead of truncating the seconds. Then on average it would be accurate.

**Comment by mathisdt** [comment]
I modified my fix for this issue, it now doesn't cut off the seconds when clocking in or out via the button (seconds are stored in the database), but when creating a report, the same algorithm is applied as in the app itself (one that doesn't take the seconds into account). This makes the app's displayed sums and the exported reports consistent, which was the original request.

By saving the seconds, there is still the possibility to create a new feature for TWT ("take seconds into account for time calculations") which would need to be done for both the app UI and for the reports and probably would be a switch in the options. But this should be a separate issue. As I stated, the sums in the UI and reports are the same now - the bug here was that the times were different (and it would be a reason to reopen this issue if this problem appears again).

Thanks to all who contributed here! I'm sorry if this isn't (yet) the solution you want, but my time is limited - I have two jobs and a family and only develop this app in my spare time.
